### PR TITLE
Add clear to list classes

### PR DESCRIPTION
**Pull request by Benestar**
This is needed as a replacement for users of the holder interfaces to remove all elements from one of the given list classes.

[Bug: T128363](https://phabricator.wikimedia.org/T128363)

### CODE REVIEW

File: src/Term/AliasGroupList.php [hunk 1]
```diff
@@ -216,4 +216,13 @@ public function toTextArray() {
 		return $array;
 	}
 
+	/**
+	 * Removes all alias groups from this list.
+	 *
+	 * @since 5.1
+	 */
+	public function clear() {
```
Comment (thiemowmde) [inline src/Term/AliasGroupList.php]: There is `Fingerprint::setAliasGroups`, so we should not need this.

File: src/Term/TermList.php [hunk 2]
```diff
@@ -190,4 +190,13 @@ public function hasTerm( Term $term ) {
 			&& $this->terms[$term->getLanguageCode()]->equals( $term );
 	}
 
+	/**
+	 * Removes all terms from this list.
+	 *
+	 * @since 5.1
+	 */
+	public function clear() {
```
Comment (thiemowmde) [inline src/Term/TermList.php]: Same here, there is `Fingerprint::setLabels` and `setDescriptions`, so this should not be needed. At least not at the moment. Did I missed a proposed use case?

Comment (Benestar) [inline src/Term/TermList.php]: Yes, when we have entities that don't contain a `Fingerprint` but only a `TermList` or an `AliasGroupList`. We definitely need the `clear` method there.

**Comment by JeroenDeDauw** [comment]
Looks good. Could perhaps also test that an empty list does not change state. But not that important

**Comment by thiemowmde** [comment]
@Benestar and I had a look in our code base. Do we really need these methods, and where? What we found:
* EntityChangeFactory clears statements, descriptions and aliases for performance reasons. This is a hack, see T113468.
* Deserializers are currently written in a way they first create an entity and then set all labels, descriptions and aliases in one step. This can be rewritten: For example, first deserialize all labels, construct a Fingerprint from that, and then construct an Item from that. No clear nor "setTermList" or something like that needed, only constructors.
* There are "patchers" that manipulate terms in a very, very complicated way. This can be refactored. A label patcher, for example, should call removeLabel, setLabel, and nothing else. It should never create a new TermList and replace the old one. It#s not a "patcher" if it does this.

**Comment by Benestar** [comment]
On hold, should not block 5.1.0

**Comment by thiemowmde** [comment]
* Do we need this in 6.0?
* Or is this the opposite of #614 and should be closed when we merge the other one?

**Comment by thiemowmde** [comment]
[What I wrote above](https://github.com/wmde/WikibaseDataModel/pull/649#issuecomment-190278907) is still true. There is still code that needs to clean these list classes. Refactoring all this code in a way that it does not need to do this is a lot of effort, and probably not really worth it.

I strongly supported #614 because it should not be possible to "repurpose" an existing entity and turn it into something completely different. There really should not be an `Item::clear`, on no entity type. But this here is very different. These are list classes. It's not really "repurposing" if code cleans a list of aliases.

Considering the odd fact that I just implemented a `clean` in https://github.com/wmde/WikibaseDataModelServices/pull/157 I suggest to merge this patch now. Pinging @brightbyte for the sake of completeness.

File: src/Statement/StatementList.php [hunk 3]
```diff
@@ -342,6 +342,15 @@ public function filter( StatementFilter $filter ) {
 	}
 
 	/**
+	 * Removes all statements from this list.
+	 *
+	 * @since 6.0
```
Comment (manicki) [inline src/Statement/StatementList.php]: this should have been 7.0 (or maybe 6.4) I guess

Comment (thiemowmde) [inline src/Statement/StatementList.php]: Thanks! Fixed in #719.
